File: Harjoitus_2/exercise2_1.py
pi = 3.141592653589793238462643383279

def init():
    print("Valitse vaihtoehto:\n1) Laske särmiön tilavuus\n2) Laske pallon tilavuus")
    try:
        userChoice = int(input("Anna valintasi: "))
    except ValueError: 
        print("Virheellinen syöte. Yritä uudelleen.")
        init()
    if userChoice == 1 or userChoice == 2:
        return userChoice
    else:
        print("Virheellinen valinta. Yritä uudelleen.")
        init()

# V = a * b * c
def sarmioKysymykset():
    a = input("Anna särmiön korkeus: ")
    b = input("Anna särmiön leveys: ")
    c = input("Anna särmiön syvyys: ")
    try:
        a = float(a)
        b = float(b)
        c = float(c)
    except ValueError:
        print("Yksi tai useampi virheellinen syöte. Yritä uudelleen.")
        return
    tilavuus = a * b * c
    print(f"Särmiön tilavuus on: {tilavuus:.2f} m³")
    return tilavuus

# V = 4/3 * π * r^3
def palloKysymykset():
    try:
        r = float(input("Anna pallon säde: "))
    except ValueError:
        print("Virheellinen syöte. Yritä uudelleen.")
        return
    tilavuus = (4/3) * pi * (r ** 3)
    print(f"Pallon tilavuus on: {tilavuus:.2f} m³")

def main():
    choice = init()
    if choice == 1:
        sarmioKysymykset()
    elif choice == 2:
        palloKysymykset()
    else:
        print("Väärä valinta. Yritä uudelleen.")
        main()

if __name__ == "__main__": 
    main()

File: cpp/Harjoitus_2/exercise2_1.cpp
#include <iostream>
#include <string>
#include <limits>
#include <cmath>
const double pi = 3.141592653589793238462643383279;

int main() {
   int uChoice;
   while (true)
   {
      std::cout << "Valitse vaihtoehto\n1) Laske särmiön tilavuus\n2) Laske pallon tilavuus\n> ";
      if (std::cin >> uChoice) { break; }
      else {
         std::cout << "Väärä syöte. Anna kokonaisluku";
         std::cin.clear();
         std::cin.ignore(std::numeric_limits<std::streamsize>::max(), '\n');
      }
   }

   if (uChoice == 1) {
      // V = a * b * c
      float a, b, c;
      while (true) {
         std::cout << "Anna särmiön korkeus:\n> ";
         if (std::cin >> a) { break; }
         else {
            std::cout << "Väärä syöte. Anna numero:\n>";
            std::cin.clear();
            std::cin.ignore(std::numeric_limits<std::streamsize>::max(), '\n');
         }
      }
      while (true) {
         std::cout << "Anna särmiön leveys:\n> ";
         if (std::cin >> b) { break; }
         else {
            std::cout << "Väärä syöte. Anna numero:\n>";
            std::cin.clear();
            std::cin.ignore(std::numeric_limits<std::streamsize>::max(), '\n');
         }
      }
      while (true) {
         std::cout << "Anna särmiön syvyys:\n> ";
         if (std::cin >> c) { break; }
         else {
            std::cout << "Väärä syöte. Anna numero:\n>";
            std::cin.clear();
            std::cin.ignore(std::numeric_limits<std::streamsize>::max(), '\n');
         }
      }
      double V = a * b * c;
      std::cout << "Tilavuus: " << V << '\n';
      return V;
   }
   if (uChoice == 2) {
      // V = 4/3 * π * r^3
      float r;
      while (true) {
         std::cout << "Anna pallon säde:\n> ";
         if (std::cin >> r) { break; }
         else {
            std::cout << "Väärä syöte. Anna numero:\n>";
            std::cin.clear();
            std::cin.ignore(std::numeric_limits<std::streamsize>::max(), '\n');
         }
      }
      double V = (4.0 / 3.0) * pi * std::pow(r,3);
      std::cout << "Tilavuus: " << V << '\n';
      return V;
   }
   else {
      std::cout << "Väärä syöte, valitse 1 tai 2";
      main();
   }
   return 0;
}
